db/version_set.cc

```diff
@@ -7137,6 +7137,20 @@ Status VersionSet::GetMetadataForFile(uint64_t number, int* filelevel,
 }
 
 void VersionSet::GetLiveFilesMetaData(std::vector<LiveFileMetaData>* metadata) {
+  if (!metadata) {
+    return;
+  }
+  assert(metadata);
+  size_t count = 0;
+  for (auto cfd : *column_family_set_) {
+    if (cfd->IsDropped() || !cfd->initialized()) {
+      continue;
+    }
```

Comment:
I've been wondering (because similar code is in the loop below) whether this is really possible or this logic came from places where we iterate `column_family_set_` without lock held.

Reply:
@ajkr  I checked the callsites for `VersionSet::GetLiveFilesMetaData` and looks like there are only two places `DBImpl::CheckConsistency` and `DBImpl::GetLiveFilesMetaData`. `DBImpl::CheckConsistency` asserts that the mutex is held. And `DBImpl::GetLiveFilesMetaData` acquires the mutex before calling `VersionSet::GetLiveFilesMetaData`.

So I guess this was added as safety check and we can try changing this to `assert(!cfd->IsDropped())` and `assert(cfd->initialized())`. Please let me know what you think.

Reply:
Sorry my thoughts on this are kind of complicated. But I am almost at the same conclusion about assertions replacing dynamic checks (or dynamic checks being corruptions rather than normal conditions, although as you mentioned already it's inconvenient that this function returns void).

Some background on `ColumnFamilySet` thread-safety: https://github.com/facebook/rocksdb/blob/3bbacda9b1b84ad202a52734ff200381608a203c/db/column_family.h#L674-L689

Here's the reason why I think `cfd->IsDropped()` should not happen in this scenario:

- `GetLiveFilesMetaData()` has DB mutex held the whole time it iterates `column_family_set_`
- No `ColumnFamilyData::SetDropped()` may run during that lock hold because calling it requires DB mutex held
- Any `ColumnFamilyData::SetDropped()` that ran before that lock hold would have removed the CFD from the `ColumnFamilySet` so iteration would not see it

For more context, here's a scenario where I think `cfd->IsDropped()` can happen:

- A caller iterates `column_family_set_` and releases DB mutex while processing each CFD
- During those processing steps, `ColumnFamilyData::SetDropped()` can happen because the DB mutex is available
- The `column_family_set_` iteration eventually reaches that dropped CFD (I think because its iterator just calls CFD::Next(), which will still reach it). Then it must explicitly skip processing that CFD by checking `cfd->IsDropped()`

Reply:
> Any ColumnFamilyData::SetDropped() that ran before that lock hold would have removed the CFD from the ColumnFamilySet so iteration would not see it

Aha, I think this is the flaw in my logic. Removing the CFD from the ColumnFamilySet is (I think) not enough. All CFHs also must be destroyed, and there could be outstanding CFHs held by the user for dropped CFDs for arbitrarily long. 

OK, so the conclusion is that this dynamic CFD::IsDropped() and not returning an error is proper. Sorry it took me a while to realize that.

Reply:
I'm still a bit confused. I want to make sure I understand this correctly.

Basically we have this `cfd->IsDropped()` check because it's possible for a CFD to be set dropped while still iterable in ColumnFamilySet, right? (due to CFH held somewhere outside)

So in `RemoveColumnFamily()`, `column_family_data_.erase(cfd_iter);` only removes the CFD from the map and won't actually call CFD's destructor as long as there's a reference to it, and because of that, it won't affect the iteration since the link for `->next` from the previous CFD is still there?

Reply:
I think your understanding is right. The thing that confused me is I assumed `ColumnFamilySet::RemoveColumnFamily()` would prevent that CF from being seen by `ColumnFamilySet` iterators. But `ColumnFamilySet` iterators don't really use `ColumnFamilySet` state; instead they rely on the CFD linked list, which can hold on to dropped CFDs much longer (until the CFD is destroyed, which requires all handles being released).